### PR TITLE
feat(nuxt3): middleware type generation improvements

### PR DESCRIPTION
**Pull request by danielroe**
### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

This PR makes a number of small changes to improve middleware DX:
* Middleware types weren't being generated on creation of middleware files - this now includes the middleware directory in the regexp for regenerating app.
* We exclude global middleware from the named middleware object (and type completion) as if it's already running, it seems needless to specify it within a page...
* We now add dir.middleware to supported v3 types
* We escape values wherever we're creating RegExps in nuxt3 templates (for example edge cases, a user might pass a nested dir or include an invalid char in their dir name)

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: packages/nuxt3/src/utils.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export const escapeRE = (str: string) => str.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&')
```
Comment (pi0) [inline packages/nuxt3/src/utils.ts:1]: Would it be overkill to use https://www.npmjs.com/package/escape-string-regexp? Seems well-tested and simple enough (https://github.com/sindresorhus/escape-string-regexp/blob/main/index.js)

Comment (danielroe) [reply packages/nuxt3/src/utils.ts:1]: Seems reasonable 👍

File: packages/nuxt3/src/pages/module.ts [hunk 2]
```diff
@@ -84,12 +91,13 @@ export default defineNuxtModule({
       async getContents () {
         const middleware = await resolveMiddleware()
         await nuxt.callHook('pages:middleware:extend', middleware)
-        const middlewareObject = Object.fromEntries(middleware.map(mw => [mw.name, `{() => import('${mw.path}')}`]))
         const globalMiddleware = middleware.filter(mw => mw.global)
+        const namedMiddleware = middleware.filter(mw => !mw.global)
+        const namedMiddlewareObject = Object.fromEntries(namedMiddleware.map(mw => [mw.name, `{() => import('${mw.path}')}`]))
```
Comment (pi0) [inline packages/nuxt3/src/pages/module.ts:96]: tip: we might use genDynamicImport in sanetize refactor

File: packages/nuxt3/src/pages/module.ts [hunk 3]
```diff
@@ -118,6 +128,7 @@ export default defineNuxtModule({
       getContents: async () => {
         const composablesFile = resolve(runtimeDir, 'composables')
         const layouts = await resolveLayouts(nuxt)
+        await nuxt.callHook('pages:layouts:extend', layouts)
```
Comment (pi0) [inline packages/nuxt3/src/pages/module.ts:131]: What about moving hook call to `resolveLayouts`?

File: packages/nuxt3/src/pages/module.ts (hunk 2 above)
Comment (danielroe) [reply packages/nuxt3/src/pages/module.ts:96]: I believe we do 😁

File: packages/nuxt3/src/pages/module.ts (hunk 3 above)
Comment (danielroe) [reply packages/nuxt3/src/pages/module.ts:131]: Good idea. And also I think there's space for improving this resolve behaviour. See here for an alternative approach: https://github.com/nuxt/framework/blob/11becef7355c9da668603a1e0129024d22a74ab1/packages/nuxt3/src/auto-imports/module.ts#L78-L90

**Review by pi0 — APPROVED** [review]

**Pull request merged**